### DISCUSSION TITLE
Derived state where a selector depends on the results of another selector.

### DISCUSSION THREAD

**Discussion by Quraian** [ngrx/platform#3273] [Q&A] [answered]
I am trying to figure out how to do something like the following:

```typescript
export const selectById = (id: string) =>
  createSelector(selectEntities, (entities) => entities.items.find((entity) => entity.id === id));

export const selectEntity = createSelector(routerSelectors.selectQueryParam('id'), selectById, (_, entity) => entity);

```

So, basically one selector depends on the result of a previous selector, how to do this in a clean way?

**Comment by markostanimirovic** [comment]
@Quraian

You can do the following:

```ts
export const selectEntityFromRoute = createSelector(
  selectEntities,
  routerSelectors.selectQueryParam('id'),
  (entities, id) => entities.find((entity) => entity.id === id)
);
```

**Reply by Quraian** [reply]
Hi @markostanimirovic 

Thanks, however, my question is more about being able to compose selectors in a more reusable way.

In your case say that I want the "find by id" as a selector on itself. So, I want `selectEntityFromRoute` selector to be composed of `routerSelectors.selectQueryParam` and `selectById` selectors, so that `selectById` is depending on the result of the other selector.

**Reply by markostanimirovic** [reply]
@Quraian 

You can create a helper function and use it as a projector for multiple selectors:

```ts
function getEntityById(entities: Entity[], id: string | undefined): Entity | undefined {
  return entities.find((entity) => entity.id === id);
}

export const selectEntityFromRoute = createSelector(
  selectEntities,
  routerSelectors.selectQueryParam('id'),
  getEntityById
);

export const selectActiveEntity = createSelector(
  selectEntities,
  selectActiveEntityId,
  getEntityById
);

// factory selector
export const selectEntityById = (id: string) => createSelector(
  selectEntities,
  (entities) => getEntityById(entities, id)
);
```

**Reply by Quraian** [reply]
Thanks @markostanimirovic 

I guess this is the closest to what I had in mind. Really appreciate your help!

**Comment by david-shortman** [comment, marked as answer]
@Quraian, By delaying the collection of the `id` param until later, you can use the result of `selectQueryParam('id")` to select an entity by id!

```typescript
export const selectByIdFunc = createSelector(
  selectEntities,
  //             👇 we return a function that takes in an id and returns the correct entity
  (entities) => (id: string) => entities.items.find((entity) => entity.id === id)
);

export const selectEntity = createSelector(
  routerSelectors.selectQueryParam('id'),
  selectByIdFunc,
  (id, getEntityById) => getEntityById(id));
```

Memoization still works properly, as `selectByIdFunc` will only change when `entities` change, and `selectEntity` will only change when the query param for "id" changes or the entities change.

**Reply by david-shortman** [reply]
Also you can create a reusable `selectById` like so:

```typescript
export const selectById = (id: string) => createSelector(
  selectByIdFunc,
  getEntityById => getEntityById(id)
);
```

**Reply by Quraian** [reply]
YES! this what I was looking for, never thought of doing it this way but I knew there must be a better way.

Thanks a lot @david-shortman !!
